### ISSUE TITLE
Clean build for homebrew

### ISSUE THREAD

**Issue by rgbkrk** [rackspace/rack#343]
This is in support of https://github.com/Homebrew/homebrew/pull/42827

Let's say we checkout the 1.0.1 branch:

```
$ git checkout 1.0.1
Note: checking out '1.0.1'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at 71a8d7c... Version 1.0.1
```

If we then build it directly, we'll end up with a `-dev` tag at the end even though we're *on the tag*:

```
$ script/build ./rack && ./rack version
Building rack!
rack version 1.0.1-dev
commit: 71a8d7c80b3652b4dd39e683bc423b8a542b0167
```

The reason for this is that we (by design) chose to append the `-dev` tag when not deploying a release via Travis. To get around it, the build can set `TRAVIS_TAG` like so:

```
$ TRAVIS_TAG=1.0.1 script/build ./rack && ./rack version
Building rack!
rack version 1.0.1
commit: 71a8d7c80b3652b4dd39e683bc423b8a542b0167
```

It might be better if we just autodetect that we're on a specific branch as well as whether we're on Travis or not.

@lvh - thanks for bearing with us and writing the homebrew recipe

**Comment by lvh** [comment]
No worries! This looks great; I think it's an improvement over relying on environment variables for specific environments like Travis or Homebrew.

Something I've done in the past is look at `git status` to see if the current checkout is identical to the currently checked out revision. E.g. if I'm on 1.0.1, make some changes & compile, 1.0.1-dev is a perfectly sensible tag. 

Unfortunately that command is well-hidden in the bowels of an org-mode file:

```
#+NAME: gitcommit
#+BEGIN_SRC emacs-lisp :exports none :tangle no
  (substring (shell-command-to-string "git rev-parse --short HEAD") 0 -1)
#+END_SRC

#+NAME: gitdesc
#+BEGIN_SRC emacs-lisp :exports none :tangle no
  (substring (shell-command-to-string "git describe") 0 -1)
#+END_SRC
```

Looks like you already have the `git rev-parse --short HEAD` part, but I think `git describe` may do what you want. Since you use unannotated tags, consider `git describe --tags` or `git describe --all` to also support upstream refs. To fully replicate the behavior, you probably also want to specify `--dirty` (or `--dirty=dev` if you want to keep the `-dev` tag), which will add a marker if the working tree isn't clean. That way you can offload all of this version number stuff to git :)

TL;DR: consider making the version number always be `git describe --all --dirty=dev`.

**Comment by rgbkrk** [comment]
> That way you can offload all of this version number stuff to git :)
> TL;DR: consider making the version number always be git describe --all --dirty=dev.

Ok, yeah that's pretty solid.

Before reading this I prepped a PR (https://github.com/lvh/homebrew/pull/2) against your original PR.

**Comment by lvh** [comment]
Okay, cool; how do you want to proceed?

**Comment by rgbkrk** [comment]
```
$ git describe --all --dirty=dev
tags/1.0.1
```

Do you mean

```
$ git describe --tags --dirty=dev
1.0.1
```

**Comment by lvh** [comment]
Oh, huh -- that's not how I remembered it worked -- I thought --all just grabbed whatever is in `refs/` and threw away all but the last part (so tags implicitly work, because `refs/tags/1.0.1` or whatever). But yes, that's a better result :-)

**Comment by rgbkrk** [comment]
If they'll take it as is, I'd love to see 1.0.1 go forward while we figure out what we're doing. Especially since this only appends the `dev` when the workspace is dirty (we kept it for denoting anything non-tagged and relied on the commit).

```
$ git describe --tags --dirty='-dev'
1.0.1-1-ge752642
$ # edit file, don't commit
$ git describe --tags --dirty='-dev'
1.0.1-1-ge752642-dev
```

**Comment by lvh** [comment]
Okay, that works. On it.

**Comment by lvh** [comment]
Following up with https://github.com/Homebrew/homebrew/pull/42827: looks like I dropped the ball on this, sorry :(

@rgbkrk: Current commit that's up for inclusion in homebrew adds the `-dev` by default. Would you like to see that one land in homebrew, or update to 1.0.2 first? FWIW I've submitted a PR :)

**Comment by rgbkrk** [comment]
I think having that land in homebrew is better than nothing. I'm seeing the same issue (`-dev` showing up) with my local builds of a different client too.

**Comment by jrperritt** [comment]
i think this landed. closing
